src/green/token.rs

```diff
@@ -1,36 +1,59 @@
-use std::sync::Arc;
+use std::{fmt, hash, mem::ManuallyDrop, ptr, sync::Arc};
 
 use crate::{cursor::SyntaxKind, SmolStr, TextUnit};
 
+#[repr(align(2))] // NB: this is an at-least annotation
 #[derive(Debug, PartialEq, Eq, Hash)]
-struct GreenTokenData {
+pub(crate) struct GreenTokenData {
     kind: SyntaxKind,
     text: SmolStr,
 }
 
 /// Leaf node in the immutable tree.
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
 pub struct GreenToken {
-    data: Arc<GreenTokenData>,
+    pub(super) ptr: ptr::NonNull<GreenTokenData>,
 }
 
+unsafe impl Send for GreenToken {} // where GreenTokenData: Send + Sync
+unsafe impl Sync for GreenToken {} // where GreenTokenData: Send + Sync
```

Comment:
Can't we just add this where bound?

Reply:
No, `GreenToken` is public and `GreenTokenData` is private.